Documentation and general cleanup for #493, #495

yjunechoe opened the pull request:
# Summary

This will be a series of (mostly) documentation-related changes complementing the two recent PRs:
- #493
- #495

Below is the (evolving) roadmap for this PR. Please feel free to interject at any point!

## 1) Chore

- [x] Remove the usage of `vars()` for column selection in docs. `vars(a)` becomes `a` and `vars(a, b)` becomes `c(a, b)`. 
- [x] Clean up capturing user-expression in `columns` in validation functions: the input should only be `enquo()`-ed once
- [x] Remove/refactor some utilities no longer used:
  - [x] `as_vars_fn()` (now `as_c_fn()`)
- Note any missing gaps in the newly introduced features and add to (3)
- Catch and fix any bugs spotted in the process
  - [x] resolving `c()`-expr column selection inside `serially()` is too eager
  - [x] `serially()` YAML section example doesn't run: "Error: There must be at least one `test_*()` function call in `serially()`."
  - [x] `col_exists()` allows all kinds of evaluation errors to go through, not just user-specified selection of non-existent columns

## 2) Documentation

- [x] Document tidyselect in `columns` param
- [x] Document (more complete) tidyselect in Column Names section
- [x] Document glue syntax in `label` param
- [x] Document multi-length vector support in `label` param

Copy pastes:

- [x] Make special `columns` param docs for `rows_*()` functions, making `everything()` the default
- [x] Edit Column Names section in individual function docs
- [x] Add Labels section in individual function docs
  - Only show the relevant a subset of glue variables for each function

## 3) Feature completeness (more of a wishlist)

- [x] refactor `get_column_text()` for writing `columns` expr to yaml
- [x] yaml functions should now write columns as `c(a)` instead of `vars(a)` if user only specifies `columns = a`
  - [x] test consumption of `columns: c(a)` and `columns: c(a, b)`
- Rest are too big/out of scope and moved to section X

## 4) Finishing touches

### News items
- [x] glue syntax in `label`
- [x] more complete tidyselect in `columns`
- [x] note beginning of deprecation process for `vars()` in `columns` (?)
- [x] encourage users to wrap the input to `columns` in `all_of()` if it's an external vector (like in `dplyr::select()`)

### Additional tests
- [x] test examples embedded in sections
- [x] ensure all tests pass
- [x] ensure pkgdown builds

## X) Bigger refactoring tasks that should be handled outside of this PR 

- yaml should default to reading/writing `columns` as expression
- Remove/refactor some utilities no longer used:
  - `uses_tidyselect()`
  - `exported_tidyselect_fns()`
- `has_columns()` currently relies on `columns = vars(...)` and could benefit from tidyselect (same situation as `col_exists()`, just more lightweight)
- `info_columns()` does not allow tidyselect expressions over column type/values like `where(is.character)` when `tbl` is loaded lazily.

# Related GitHub Issues and PRs

- Ref: #493, #495

# Checklist

- [x] I understand and agree to the [Code of Conduct](https://www.contributor-covenant.org/version/2/1/code_of_conduct.html).
- [x] I have listed any major changes in the [NEWS](https://github.com/rstudio/pointblank/blob/main/NEWS.md).
- [x] I have added [`testthat`](https://github.com/r-lib/testthat) unit tests to [`tests/testthat`](https://github.com/rstudio/pointblank/tree/main/tests/testthat) for any new functionality.

yjunechoe commented:
Hi @rich-iannone - not urgent at all, but when you get a chance could you quickly course correct / review the current draft of the following function doc edits? The changes are currently only made in `col_vals_gt.R` - I'm planning to make variants of these for copy-paste/inheritParams.

If current draft is wordy, I can move some stuff out into NEWS (like the status of `vars()` and reminding users to use `all_of()`).

Thanks!

---

## 1) `columns` argument

### Previous

https://github.com/rstudio/pointblank/blob/ca26d12fb31e781ff5228be31fc4458fe36774b3/R/col_vals_gt.R#L54-L59

### Current

```r
#' @param columns *The target columns*
#' 
#'   `<tidy-select>` // **required**
#' 
#'   A column-selecting expression, as one would use inside `dplyr::select()`.
#'   Specifies the column (or a set of columns) to which this validation should
#'   be applied. See the *Column Names* section for more information.
```

---

## 2) "Column Names" section

### Previous

https://github.com/rstudio/pointblank/blob/ca26d12fb31e781ff5228be31fc4458fe36774b3/R/col_vals_gt.R#L194-L201

### Current

```r
#' @section Column Names:
#' 
#' `columns` may be a single column (as symbol `a` or string `"a"`) or a vector
#' of columns (`c(a, b, c)` or `c("a", "b", "c")`). `{tidyselect}` helpers
#' are also supported, such as `contains("date")` and `where(is.double)`. If
#' passing an *external vector* of columns, it should be wrapped in `all_of()`.
#' 
#' When multiple columns are selected by `columns`, the result will be an
#' expansion of validation steps to that number of columns (e.g.,
#' `c(col_a, col_b)` will result in the entry of two validation steps).
#' 
#' Previously, columns could be specified in `vars()`. This continues to work, 
#' but `c()` offers the same capability and supersedes `vars()` in `columns`.
```

---

## 3) `label` argument

### Previous

https://github.com/rstudio/pointblank/blob/ca26d12fb31e781ff5228be31fc4458fe36774b3/R/col_vals_gt.R#L134-L139

### Current

```r
#' @param label *Optional label for the validation step*
#' 
#'   `vector<character>` // *default:* `NULL` (`optional`)
#' 
#'   Optional label for the validation step. This label appears in the *agent*
#'   report and, for the best appearance, it should be kept quite short. See
#'   the *Labels* section for more information.
```

---

## 4) "Labels" section

### Previous

Didn't exist. This is section is new.

### Current

```r
#' @section Labels:
#' 
#' `label` may be a single string or a character vector that matches the number
#' of expanded steps. `label` also supports `{glue}` syntax and exposes the
#' following dynamic variables contextualized to the current step:
#'   
#' - `"{.step}"`: The validation step name
#' - `"{.col}"`: The current column name
#' - `"{.seg_col}"`: The current segment's column name
#' - `"{.seg_val}"`: The current segment's value/group
#'     
#' The glue context also supports ordinary expressions for further flexibility
#' (ex: `"{toupper(.step)}"`) as long as they return a length-1 string.
```

Note that the exact list items will vary depending on whether the function indeed accepts a `columns` and/or `segments` argument.

rich-iannone commented:
This is great! On top of this, I wonder if functions like `rows_distinct()` and `rows_complete()` should default to `everything()` (but continue to work with `NULL`, in case somebody explicitly sets that)? Only because those functions act the opposite way by considering all the columns. Let me know what you think because the concept change (and lack of good documentation for `columns` in those functions) has been a point of confusion for a number of users.

yjunechoe commented:
Thanks! I'll push forward with copy pastes in that case.

> On top of this, I wonder if functions like rows_distinct() and rows_complete() should default to columns = everything() (but continue to work with NULL, in case somebody explicitly sets that)?

Makes sense! I implemented that change partly in #493, where the default is still `NULL` but that gets translated to `everything()` internally. Here I can just move `everything()` up into the function formals for clarity.

https://github.com/rstudio/pointblank/blob/ca26d12fb31e781ff5228be31fc4458fe36774b3/R/rows_distinct.R#L280-L284

rich-iannone commented:
I think the skipping behavior with `columns = NULL` on the `col_vals_*()`/`col_is_*()` functions is new. I tried setting `columns = NULL` with `main` (before the 2 PRs) and got an ugly error, which is undocumented and shouldn't really happen when using an agent.

If you're up for it, this would be another chance for modernization of the API to use more tidyselect semantics. Right now in `main`, `columns = NULL` results in a single, skipped validation step. That's better than before but I'm not certain it's what we want. We have an `active` argument that is better meant for that, where we could supply a logical or a function that should resolve to a logical (which I believe is in the same/similar scope as a function invoked as `preconditions`; the helper function is `has_columns()` and this is the PR that introduces it: https://github.com/rstudio/pointblank/pull/240). This work (should you choose to accept it!) can all be deferred to a non-documentation PR.

Here are some examples, and I'll summarize what happens with `main`:

```r 
create_agent(exibble) |>
  col_vals_lt(columns = NULL, value = 5) |>
  interrogate()
```

^ skips. Probably not ideal, you get a weird autobrief (`"Expect that values in NA (computed column) ... "`)

```r
create_agent(exibble) |>
  col_vals_lt(columns = starts_with("z"), value = 5) |>
  interrogate()
```

^ same result as before but more likely for the user to hit. It can be dangerous safety-wise to use tidyselect because the match can be arbitrary. Users should be advised to pair with a `has_columns()` check in the `active` argument.

```r
create_agent(exibble) |>
  col_vals_lt(columns = num,  value = 5, active = ~ . %>% has_columns(num)) |>
  interrogate()
```

^ this ought to work but tidyselect is not yet ported to `active`/`has_columns()`. So we get a console error with `interrogate()` (the error is not captured internally); we don't get a full validation report.

```r
create_agent(exibble) |>
  col_vals_lt(columns = num, value = 5, active = ~ . %>% has_columns("num")) |>
  interrogate()
```
```r
create_agent(exibble) |>
  col_vals_lt(columns = num, value = 5, active = ~ . %>% has_columns(vars(num))) |>
  interrogate()
```

^ The above two cases (with string-based column names or with `vars()`) work as before.

```r
create_agent(exibble) |>
  col_vals_lt(columns = starts_with("z"), value = 5, active = ~ . %>% has_columns(starts_with("z"))) |>
  interrogate()
```

^ doesn't work, but closer to a safety net that is useful.

This is a lot to sort though, so let me know if more work in this direction is worthwhile!

yjunechoe commented:
Thanks for the details comments on `columns = NULL`! I've moved it over to #497 to tackle this separately since it (thankfully) doesn't derail the current direction of this PR.

I've also now made `everything()` visible as the default value of `columns` for `rows_*()` functions.

```r
args(rows_distinct)
#> function (x, columns = tidyselect::everything(), preconditions = NULL, 
#>     segments = NULL, actions = NULL, step_id = NULL, label = NULL, 
#>     brief = NULL, active = TRUE) 
```

Back to copy-pasting!

rich-iannone commented:
It's really a lot of files! I sometimes regret that choice.

yjunechoe commented:
I think this more or less covers the necessary doc changes w.r.t. `columns` and `label`! I've also removed references to `vars()` in `columns` and replaced them with `c()` (or just mention tidyselect) as much as possible. I'll follow up on the remaining refactoring/bugfix tasks elsewhere.

(Also - any clue about the error in codecov GHA? It runs fine locally and I can't seem to pinpoint the failing test)

rich-iannone reviewed (approved):
LGTM!!

rich-iannone commented:
That was a huge amount of work but these are super good changes!

I watched the codecov workflow run and couldn’t get any info about why the tests fail in that workflow specifically. It’s something that could be looked at later. 

As for this PR, all approved. Feel free to merge this in at your leisure!

yjunechoe commented:
Thanks!

And RE: codecov, I saw one of the previous fails leaking a possible hint: 

https://github.com/rstudio/pointblank/actions/runs/6685660721/job/18164187556

> Error: Error: Failure in `/tmp/Rtmp16024Y/R_LIBS177c5e9c956b/pointblank/pointblank-tests/testthat.Rout.fail`
TRUE) or duckdb::duckdb_shutdown(drv) to avoid this. 

This was surprising because my impression was that we skip DB related tests on CI. But I'll see if that clue leads us to a fix.

And since I unfortunately cannot reproduce this myself, I'm going to try to debug this by hitting the action directly in another PR - just a heads up that this might get noisy!
